packages/aws_signature_v4/lib/src/request/canonical_request/canonical_request.dart

```diff
@@ -0,0 +1,268 @@
+      'Expiration must be greater than 1 second and less than 7 days',
+    );
+
+    final canonicalHeaders = CanonicalHeaders(headers);
+    final signedHeaders = SignedHeaders(canonicalHeaders);
+    serviceConfiguration.applyPresigned(
+      queryParameters,
+      request: request,
+      credentialScope: credentialScope,
+      algorithm: algorithm,
+      expiresIn: expiresIn.inSeconds,
+      signedHeaders: signedHeaders,
+      credentials: credentials,
+    );
+
+    return CanonicalRequest._(
+      request: request,
+      queryParameters: queryParameters,
+      canonicalQueryParameters: CanonicalQueryParameters(queryParameters),
+      headers: headers,
+      canonicalHeaders: canonicalHeaders,
+      signedHeaders: signedHeaders,
+      credentialScope: credentialScope,
+      algorithm: algorithm,
+      expiresIn: expiresIn,
+      presignedUrl: true,
+      contentLength: contentLength,
+      payloadHash: payloadHash,
+      serviceConfiguration: serviceConfiguration,
+    );
+  }
+
+  /// {@macro aws_signature_v4.canonical_request}
+  CanonicalRequest._({
+    required this.request,
+    required this.queryParameters,
+    required this.canonicalQueryParameters,
+    required this.headers,
+    required this.canonicalHeaders,
+    required this.signedHeaders,
+    required this.credentialScope,
+    required this.contentLength,
+    required this.payloadHash,
+    required this.serviceConfiguration,
+    required this.presignedUrl,
+    this.algorithm,
+    Duration? expiresIn,
+  })  : normalizePath = serviceConfiguration.normalizePath,
+        omitSessionTokenFromSigning = serviceConfiguration.omitSessionToken,
+        expiresIn = expiresIn?.inSeconds;
+
+  /// Returns the normalized path with double-encoded path segments.
+  ///
+  /// Uses [url] to normalize the path.
+  static String _canonicalPath(
+    AWSBaseHttpRequest request, {
+    required bool normalizePath,
+  }) {
+    var path = normalizePath ? url.normalize(request.path) : request.path;
+
+    // `normalize` removes leading and trailing slashes which should be preserved.
+    if (normalizePath) {
+      if (request.path.endsWith('/')) {
+        path = path.ensureEndsWith('/');
+      }
+      if (!request.path.startsWith('/')) {
+        path = path.ensureStartsWith('/');
+      }
+    }
+
+    return path.split('/').map(Uri.encodeComponent).join('/');
+  }
+
+  /// Creates the canonical request string.
+  @override
+  String toString() {
+    final sb = StringBuffer();
+    sb.writeln(request.method.value);
+    sb.writeln(canonicalPath);
+    sb.writeln(canonicalQueryParameters);
+    sb.writeln(canonicalHeaders);
+    sb.writeln(signedHeaders);
+    sb.write(payloadHash);
+    return sb.toString();
+  }
+}
+
+extension on String {
+  String ensureStartsWith(String s) {
+    if (!startsWith(s)) {
+      return '$s$this';
+    }
+    return this;
+  }
+
+  String ensureEndsWith(String s) {
+    if (!endsWith(s)) {
+      return '${this}$s';
```

Comment:
No since `$` is valid in variable names, removing it would confuse between two variables, `this` and `this$s`

Reply:
Hmm, if I'm not mistaken, I think to specify you want `this$s` instead of `this` you would have to do:
```dart
'${this$s}'
```
Otherwise line 257 would be broken right?

Reply:
I mean you can try in DartPad lol